src/query/src/dist_plan/analyzer.rs

```diff
@@ -190,12 +196,20 @@ impl PlanRewriter {
                     self.stage.push(plan)
                 }
             }
-            Commutativity::TransformedCommutative(transformer) => {
+            Commutativity::TransformedCommutative {
+                transformer,
+                expand_on_parent,
+            } => {
                 if let Some(transformer) = transformer
-                    && let Some(plan) = transformer(plan)
+                    && let Some(changed_plan) = transformer(plan)
                 {
-                    self.update_column_requirements(&plan);
-                    self.stage.push(plan)
+                    debug!("PlanRewriter: transformed plan: {changed_plan:#?} from {plan}");
+                    if let Some(last_stage) = changed_plan.last() {
+                        // update the column requirements from the last stage
+                        self.update_column_requirements(last_stage);
+                    }
```

Comment:
Should we update column requirements for all changed plans instead of only the last one?

Reply:
> Should we update column requirements for all changed plans instead of only the last one?

seems not needed? since only the last stage need to have all those column requirements, so merge scan can bring up those required columns